src/__experimental__/components/decimal/decimal.component.js

```diff
@@ -1,130 +1,255 @@
-    const seperator = `\\${format.separator}`;
-    const validDecimalMatcher = new RegExp(`^[-]?[\\d${delimiter}]*[${seperator}]?\\d{0,${precision}}?$`);
-
+    const separator = `\\${format.separator}`;
+    const validDecimalMatcher = new RegExp(`^[-]?[\\d${delimiter}]*[${separator}]?\\d{0,${precision}}?$`);
     return validDecimalMatcher.test(value);
   }
 
+  createEvent = () => {
+    const standardVisible = this.toStandardDecimal(this.state.visibleValue);
+    const formattedValue = this.isNaN(standardVisible) ? this.state.visibleValue : this.formatValue(standardVisible);
+    return {
+      target: {
+        name: this.props.name,
+        id: this.props.id,
+        value: {
+          rawValue: this.state.value,
+          formattedValue
+        }
+      }
+    };
+  }
+
   onChange = (ev) => {
-    const { target } = ev;
-    const { value, selectionEnd } = ev.target;
+    const { target: { value } } = ev;
+    this.setState({ value: this.toStandardDecimal(value), visibleValue: value });
+  }
+
+  onPaste = (ev) => {
+    const value = this.replace(ev);
     const isValid = this.isValidDecimal(value);
+    if (!isValid) {
+      ev.preventDefault();
+    }
+  }
 
-    if (isValid) {
-      if (!this.isControlled) {
-        this.setState({ value });
+  callOnChange = () => {
+    if (this.props.onChange) {
+      this.props.onChange(this.createEvent());
+    }
+  }
+
+  /**
+   * Perform a string replacement on the input so we can see what the value will be after the change
+   * This allows us to prevent the user input if it is invalid
+   */
+  replace = (ev) => {
+    const { target: { selectionStart, selectionEnd, value }, clipboardData, type } = ev;
+    let { key: change } = ev;
+
+    if (type === 'paste') {
+      change = clipboardData.getData('text/plain');
+    }
+
+    if (Events.isDeletingKey(ev)) {
+      change = '';
+    }
+    return (
+      value.substring(0, selectionStart) + change + value.substring(selectionEnd)
+    );
+  };
+
+  isValidKeyPress = (ev) => {
+    const { delimiter, separator } = I18nHelper.format();
+
+    if (Events.isNumberKey(ev) || ev.key === delimiter || ev.key === separator || Events.isDeletingKey(ev)) {
+      return this.isValidDecimal(this.replace(ev));
+    }
+
+    if (ev.ctrlKey) {
+      // user is doing some browser related task, we don't want to interfere with that.
+      // we do want to check paste, we do that onPaste
+      return true;
+    }
+
+    if (Events.isMinusKey(ev) && ev.target.selectionStart === 0) {
+      // user is entering a negative symbol at the start of the number
+      return true;
+    }
+
+    return false;
+  }
+
+  onKeyPress = (ev) => {
+    // We're checking to see if the proposed change is valid, if not we block the user input
+    if (!this.isValidKeyPress(ev)) {
+      ev.preventDefault();
+    }
+    if (this.props.onKeyPress) {
+      this.props.onKeyPress(ev);
+    }
+  }
+
+  componentDidUpdate(prevProps, prevState) {
```

Comment:
Not really a blocking comment but I think it would be nice if we'd all follow airbnb's recommendation on ordering of lifecycle methods. It makes components definitely more readable and easier to grasp.
https://github.com/airbnb/javascript/tree/master/react#ordering
Could be also enforced by an ESLint rule.